p2panda-net/src/network.rs

```diff
@@ -0,0 +1,647 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+
+use std::net::SocketAddr;
+use std::sync::Arc;
+use std::time::Duration;
+
+use anyhow::{anyhow, Context, Result};
+use futures_lite::StreamExt;
+use iroh_gossip::net::{Gossip, GOSSIP_ALPN};
+use iroh_gossip::proto::Config as GossipConfig;
+use iroh_net::endpoint::TransportConfig;
+use iroh_net::key::SecretKey;
+use iroh_net::relay::{RelayMap, RelayNode};
+use iroh_net::util::SharedAbortingJoinHandle;
+use iroh_net::{Endpoint, NodeAddr, NodeId};
+use p2panda_core::{PrivateKey, PublicKey};
+use tokio::sync::{broadcast, mpsc};
+use tokio::task::JoinSet;
+use tokio_util::sync::CancellationToken;
+use tracing::{debug, error, error_span, warn, Instrument};
+
+use crate::addrs::DEFAULT_STUN_PORT;
+use crate::config::{Config, DEFAULT_BIND_PORT};
+use crate::discovery::{Discovery, DiscoveryMap};
+use crate::engine::Engine;
+use crate::handshake::{Handshake, HANDSHAKE_ALPN};
+use crate::protocols::{ProtocolHandler, ProtocolMap};
+use crate::{NetworkId, RelayUrl, TopicId};
+
+/// Maximum number of streams accepted on a QUIC connection.
+const MAX_STREAMS: u32 = 1024;
+
+/// Maximum number of parallel QUIC connections.
+const MAX_CONNECTIONS: u32 = 1024;
+
+/// Timeout duration for discovery of at least one peer's direct address.
+const ENDPOINT_WAIT: Duration = Duration::from_secs(5);
+
+#[derive(Debug, PartialEq)]
+pub enum RelayMode {
+    Disabled,
+    Custom(RelayNode),
+}
+
+/// Creates an overlay network for peers grouped under the same network identifier.
+///
+/// All peers can subscribe to multiple topics in this overlay and hook into a data stream per
+/// topic where they'll send and receive data.
+#[derive(Debug)]
+pub struct NetworkBuilder {
+    bind_port: Option<u16>,
+    direct_node_addresses: Vec<NodeAddr>,
```

Comment:
We talked about renaming this field (in many other places as well) to something like `peer_addresses` or `known_peer_addresses` or `direct_peer_addresses` to avoid confusion with _our own_ direct addresses. Doesn't need changing here in this PR, leaving a comment so we can consider it further though.